src/validation.cpp

```diff
@@ -2854,6 +2897,143 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+                if (int(height) != pindex->nHeight) {
+                    return false;
+                }
+                for (uint32_t i = 0; i < data.batchCount; i++) {
+                    auto batch = cache.GetAuctionBatch(vaultId, i);
+                    assert(batch);
+                    if (auto bid = cache.GetAuctionBid(vaultId, i)) {
+                        auto amountToFill = DivideAmounts(bid->second.nValue, COIN + data.liquidationPenalty);
+                        auto amountToBurn = bid->second.nValue - amountToFill;
+                        if (amountToBurn > 0) {
+                            cache.AddBalance(chainparams.GetConsensus().burnAddress, {bid->second.nTokenId, amountToBurn});
+                            CPoolSwapMessage obj;
+                            obj.from = chainparams.GetConsensus().burnAddress;
+                            obj.to = chainparams.GetConsensus().burnAddress;
+                            obj.idTokenFrom = bid->second.nTokenId;
+                            obj.idTokenTo = DCT_ID{0};
+                            obj.amountFrom = amountToBurn;
+                            obj.maxPrice = POOLPRICE_MAX;
+                            auto poolSwap = CPoolSwap(obj, pindex->nHeight);
+                            // swap tokenID -> USD -> DFI
+                            auto token = cache.GetToken("USD");
+                            assert(token);
+                            poolSwap.ExecuteSwap(cache, {bid->second.nTokenId, token->first, DCT_ID{0}});
+                        }
+                        cache.CalculateOwnerRewards(bid->first, pindex->nHeight);
+                        for (const auto& col : batch->collaterals.balances) {
+                            cache.AddBalance(bid->first, {col.first, col.second});
+                        }
+                        // return rest loan to vault if any
+                        amountToFill -= batch->loanAmount.nValue;
+                        if (amountToFill > 0) {
+                            cache.AddLoanToken(vaultId, {batch->loanAmount.nTokenId, amountToFill});
+                        }
+                    } else {
+                        cache.AddLoanToken(vaultId, batch->loanAmount);
+                        for (const auto& col : batch->collaterals.balances) {
+                            cache.AddVaultCollateral(vaultId, {col.first, col.second});
+                        }
+                    }
+                }
+                auto vault = cache.GetVault(vaultId);
+                assert(vault);
+                vault.val->isUnderLiquidation = false;
+                cache.StoreVault(vaultId, *vault.val);
+                cache.EraseAuction(vaultId, pindex->nHeight);
+                cache.TransferVaultInterest(vaultId, pindex->nHeight, vault.val->schemeId, {});
+                return true;
+            }, pindex->nHeight);
+        }
```

Comment:
Can we move this section of code to a function, function `CChainState::ConnectBlock` is super long already

Reply:
We will move composite dex burning in new function. All of these iterations can be in its own function, @ShengguangXiao if you're intested split them into distinct functions.